### ISSUE TITLE
Tabbing outside the modal can cause the page to scroll

### ISSUE THREAD

**Issue by supachailllpay** [twbs/bootstrap#19642]
When I continue press tab until the last tab pressed then it seems go out of modal.

Example:
https://jsfiddle.net/7p330rgs/

I look at the code about enforceFocus function. I think one of easiest way to do is create the dummy focus (last focus on modal) if this dummy focused then force focus back to modal. BTW I don't have any test about it.

Thank you.

**Comment by cvrebert** [comment]
@supachailllpay What browser and OS are you using?

**Comment by cvrebert** [comment]
Hmm, on OS X, this only reproduced for me on Chrome. And the link never gets fully focused such that the user could interact with it, but it does get scrolled into view.
Anyone test this on Windows?

**Comment by patrickhlauke** [comment]
using https://jsfiddle.net/w7o3bmxe/3/ i can confirm that on Windows in Chrome, Firefox, IE, Edge the link receives focus (but cannot be activated)

**Comment by cvrebert** [comment]
> the link receives focus

Well, FWICT that seems to be unavoidable, assuming I understand HTML's focus model correctly; none of the focus-related events are cancelable (WTF spec writers...). So what matters is whether the focus is immediately shifted back to the modal via an event handler, such that the user cannot possibly interact with the link.

Noodling about it more, I think the bug here is that we ought to record (when opening the modal) and restore the scroll position of the `<body>` when we refocus the modal. Probably we haven't noticed this bug before since it's a fairly rare case: the next focusable element after the modal is offscreen, thus causing a scroll, which doesn't get counteracted when we refocus the modal. Most pages have enough focusable elements that this situation doesn't occur.

**Comment by patrickhlauke** [comment]
Longer term, integrating something like ally.js may be worth considering - see http://allyjs.io/tutorials/accessible-dialog.html

**Comment by rodneyrehm** [comment]
[Focusing algorithm is missing a "scroll the element into view" step](https://github.com/whatwg/html/issues/94) is the issue trying to get the scroll-on-focus thing to be specified. 

Since the scoll happens on `focus`, it would've been possible to redirect focus upon `focusin`, if browsers followed the [DOM Level3 FocusEvent order](https://www.w3.org/TR/DOM-Level-3-Events/#h-legacy-focusevent-event-order) - which they *don't*.

The only way to *prevent* this from happening is to prevent the `focus` itself. This can be achieved by

* intercepting the <kbd>Tab</kbd> and <kbd>Shift Tab</kbd> keys ([`ally.maintain.tabTocus`](http://allyjs.io/api/maintain/tab-focus.html))
* removing all elements outside of the dialog from the document's tabbing order by setting them to `tabindex="-1"` ([`ally.maintain.disabled`](http://allyjs.io/api/maintain/disabled.html))

You can *recover* from the scroll-on-focus by observing the document's scroll-position on `keydown` and resetting on `blur` on the element outside of the dialog. Not a nice approach, imho.

**Comment by mdo** [comment]
[**Bootstrap 3 is no longer being officially developed or supported.**](https://github.com/twbs/bootstrap/issues/20631)

All work has moved onto our next major release, v4. As such, this issue or pull request is being closed as a "won't fix." For additional help and support, we recommend [utilizing our community resources](https://github.com/twbs/bootstrap/blob/master/README.md#community). Thanks for your understanding, and see you on the other side of v4!

<3,
@mdo and team
